adding class file parsing module

jshlbrd opened the pull request:
This PR adds a JAR class file parsing module. The module relies on python-javatools and parses the constant pool table and the classes provided by and required by the class file. Thanks!

jshlbrd committed:
adding class file parsing module

jshlbrd commented:
Sample output can be seen here: https://github.com/jshlbrd/laikaboss-modules/tree/master/meta_class

jshlbrd committed:
removed logging import

kglm commented:
Nice and simple, beautiful module. Got one recommendation and two suggestions for you.

Recommend: You add the name JAVA to the module name. This provides more context at the global scope for the type of class being meta'd.

Suggest: You can import and except ScanError instead of having to do all three framework exceptions (QuitScanException, GlobalScanTimeoutError, GlobalModuleTimeoutError). These exceptions inherit ScanError, so ScanError handles them all. This was an addition after several modules were developed, betting you just copy-pasted from one of these old ones (as you should, don't reinvent the wheel).

Suggest: We have tried to stick with 4 space indentations. I noticed the scan_module line in the Yara dispatch rule uses a tab.

marnao commented on laikaboss/modules/meta_class.py:
```diff
@@ -0,0 +1,44 @@
+        moduleResult = []
+
+        try:
+                class_obj = unpack_class(scanObject.buffer)
```
You've got inconsistent indenting here and in the except block. Please use 4 spaces for all indenting.

jshlbrd committed:
Changed name of module
Commented out java class requires metadata collection
Fixed spacing issues

jshlbrd committed:
Fixed dispatch.yara

jshlbrd commented:
Thanks for the feedback @marnao and @kglm. I think I've addressed all of your feedback in these updates. 

One thing to note, I used a new set of JAR files for testing these changes and noticed that python-javatools throws a lot of errors when trying to parse the class requires data. The errors also appear when running the native python-javatools scripts, so it isn't a fault of this module-- in any case, I've commented out the two lines that do the class requires data collection. I'll try to reach out to the author of python-javatools and see if we can address this later. In the mean time, most users will probably find the class provides data and the constants pool data plenty useful.

Let me know if anything is needed on this module. Thanks!

kglm commented:
As an alternative, you could consider wrapping those two lines within a try-except (if the errors thrown are easily enumerable) and just call `pass` in the except block. That way, if it does work, you benefit with the metadata and if it errors, the module cleanly skips over that piece. Maybe even add a metadata field noting the error instead of just passing.

jshlbrd committed:
Wrapped class_requires in try/except

jshlbrd commented:
That was a great suggestion-- I put that code in a try-except and the exception adds this flag to the object: 'java_class:err:class_requires'. Tried to keep this in line with other error flags, and since this particular error is generic, I don't think a detailed error message is necessary.

kglm commented:
One last detail, add the except ScanError: raise to the inner try-except and this is ready to go.

jshlbrd committed:
Updated exceptions

jshlbrd commented:
Good timing, I took another look at python javatools and it looks like there are a couple exceptions that can be thrown throughout the unpacking process ... so now they're handled the way they probably should've been from the beginning. :)
